src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/KeyFrameEvaluator.cs

```diff
@@ -193,8 +197,44 @@ internal bool MatchedHelper(bool firstSegment, int segmentNumber, bool botAction
             }
 
             // process each list.. start with the ones for this tier
-
             // start cv stuff first.. it runs async anyway, but the sooner we can start it the better
+            if (cvObjectsToMatch.Count > 0)
+            {
+                var cvObjectDetectionResults = CVObjectDetectionEvaluator.Matched(segmentNumber, cvObjectsToMatch);
+                if (cvObjectDetectionResults == null)
+                {
+                    // cvText results will be null until we get the async response back
+                    if (andOr == BooleanCriteria.And)
+                    {
+                        _newUnmatchedCriteria.Add("Waiting for CV Object Detection evaluation results ...");
+                        return false;
+                    }
+                }
+                else
+                {
+                    var cvObjectDetectionResultsCoount = cvObjectDetectionResults.Count;
+                    for (var i = 0; i < cvObjectDetectionResultsCoount; i++)
+                    {
+                        var resultEntry = cvObjectDetectionResults[i];
+                        if (resultEntry == null)
+                        {
+                            if (andOr == BooleanCriteria.Or)
+                            {
+                                return true;
+                            }
+                        }
+                        else
+                        {
+                            if (andOr == BooleanCriteria.And)
+                            {
+                                _newUnmatchedCriteria.Add(resultEntry);
+                                return false;
+                            }
+                        }
+                    }
+                }
+            }
```

Comment:
Assuming this is basically a copy-paste of the logic for the other CV stuff? If so, 👍

Reply:
This is a copy paste. I don't fully understand this yet, but I am getting there.